src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -0,0 +1,302 @@
+/**
+ * @file methods/ann/layer/multihead_attention.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Definition of the MultiheadAttention class.
+ *
+ * @code
+ * @article{NIPS'17,
+ *   author  = {Ashish Vaswani, Llion Jones, Noam Shazeer, Niki Parmar,
+ *              Aidan N. Gomez, Jakob Uszkoreit, Łukasz Kaiser,
+ *              Illia Polosukhin},
+ *   title   = {Attention Is All You Need},
+ *   year    = {2017},
+ *   url     = {http://arxiv.org/abs/1706.03762v5}
+ * }
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_HPP
+#define MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/layer/softmax.hpp>
+#include <mlpack/methods/ann/layer/dropout.hpp>
+#include <mlpack/methods/ann/init_rules/glorot_init.hpp>
+#include <mlpack/methods/ann/regularizer/no_regularizer.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Multihead Attention allows the model to jointly attend to information from
+ * different representation subspaces at different positions. With a single
+ * attention head, averaging inhibits this. [arxiv.org:1706.03762v5]
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam RegularizerType Type of the regularizer to be used.
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat,
+    typename RegularizerType = NoRegularizer
+>
+class MultiheadAttention
+{
+ public:
+  /**
+   * Default constructor.
+   */
+  MultiheadAttention();
+
+  /**
+   * Create the MultiheadAttention object using the specified modules.
+   *
+   * @param embedDim Total dimension of the model.
+   * @param nHeads Number of parallel attention heads.
+   */
+  MultiheadAttention(const size_t embedDim,
+                     const size_t nHeads);
```

Comment:
I suggest to replace `nHeads` to `numHeads`. I think it sounds better.
```suggestion
   * @param numHeads Number of parallel attention heads.
   */
  MultiheadAttention(const size_t embedDim,
                     const size_t numHeads);
```